pkg/kubelet/volume/populator/desired_state_of_world_populator.go

```diff
@@ -120,6 +124,32 @@ func (dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods() {
 	for _, volumeToMount := range dswp.desiredStateOfWorld.GetVolumesToMount() {
 		if _, podExists :=
 			dswp.podManager.GetPodByUID(volumeToMount.Pod.UID); !podExists {
+			// Once a pod has been deleted from kubelet pod manager, do not
+			// delete it from volume manager. Instead call out to the kubelet
+			// containerRuntime interface to verify all containers in the pod
+			// have been terminated.
+
+			podStatus, podStatusErr := dswp.kubeContainerRuntime.GetPodStatus(
+				volumeToMount.Pod.UID,
+				volumeToMount.Pod.Name,
+				volumeToMount.Pod.Namespace)
+			if podStatusErr != nil {
+				glog.Errorf(
+					"kubeContainerRuntime.GetPodStatus returned error %v. Assuming pod %q is deleted.",
+					podStatusErr,
+					volumeToMount.PodName)
+			} else {
+				for _, containerStatus := range podStatus.ContainerStatuses {
+					if containerStatus.State != kubecontainer.ContainerStateExited {
```

Comment:
Wasn't sure if it was ok for us to assume state unknown was safe to assume not running. But sure, will make that change.